### PR TITLE
Size_t-ify BIO_write() and BIO_read()

### PR DESCRIPTION
**Pull request by mattcaswell**
This creates new BIO_write_ex() and BIO_read_ex() functions that properly handle size_t input arguments and a size_t for the number of bytes read/written.

This is a two part PR. The next part will size_t-ify libssl. Once they are both in there is a small amount of additional work to do to join the two halves together.

### CODE REVIEW

File: doc/crypto/BIO_read.pod [hunk 1]
```diff
@@ -31,7 +42,10 @@ BIO_puts() attempts to write a NUL-terminated string B<buf> to BIO B<b>.
 
 =head1 RETURN VALUES
 
-All these functions return either the amount of data successfully read or
+BIO_read_ex() and BIO_write_ex() return 1 if data was successfully read or
+written, and 0 otherwise.
```
Comment (levitte) [inline doc/crypto/BIO_read.pod]: Errrr....  seeing commit 26f321440c707d89cf97903668060fca637ed44d, this can't be true.

Comment (mattcaswell) [inline doc/crypto/BIO_read.pod]: Oh. Yes. I'm confusing myself with the SSL_read_ex() and SSL_write_ex() behaviour.

File: crypto/bio/bio_meth.c [hunk 2]
```diff
@@ -63,12 +63,45 @@ int BIO_meth_set_write(BIO_METHOD *biom,
...
     return biom->bread;
 }
 
+/* Conversion for old style bread to new style */
+int bread_conv(BIO *bio, char *out, size_t outl, size_t *read)
+{
+    int ret;
+
+    if (outl > INT_MAX)
+        return 0;
```
Comment (levitte) [inline crypto/bio/bio_meth.c]: This seems wrong.  Instead if returning 0, I'd set `outl = INT_MAX;` here

File: doc/crypto/BIO_read.pod (hunk 1 above)
Comment (mattcaswell) [inline doc/crypto/BIO_read.pod]: I would prefer it that we just returned 0 on failure and 1 on success...and that is what SSL_read_ex() and SSL_write_ex() do. Perhaps I should change it. Mostly applications don't care what the value of the return is...a failure is a failure...if you want more info look at the error queue.

File: crypto/bio/bio_meth.c [hunk 3]
```diff
@@ -63,12 +63,45 @@ int BIO_meth_set_write(BIO_METHOD *biom,
...
+
+    ret = bio->method->bread_old(bio, out, (int)outl);
+
+    if (ret <= 0) {
+        *read = 0;
+        return ret;
+    }
```
Comment (levitte) [inline crypto/bio/bio_meth.c]: If `ret == 0` and this is a non-blocking BIO, this is incorrectly handled.  I'd simply change `if (ret <= 0)` to `if (ret < 0)`

Comment (mattcaswell) [inline crypto/bio/bio_meth.c]: Hmmm...I'm not sure I understand. If `ret == 0` then surely we need to return 0 (it's a failure). Changing to `if ( ret < 0)` would mean we would return 1 in this instance.

Comment (levitte) [inline crypto/bio/bio_meth.c]: Doesn't `return == 0` mean "there was zero bytes of data available to read at this moment / location"? I wouldn't necessarily call that a failure, and most definitely not on a non-smoking BIO.

Comment (levitte) [inline crypto/bio/bio_meth.c]: That was supposed to be "non-blocking"...

Comment (mattcaswell) [inline crypto/bio/bio_meth.c]: If there is data to be read prior to reaching eof then as much data as we can get will be read into the buffer, and then a success return will be given. If you attempt to read again then there is no data to be read and it returns as a failure. This is the same as fread, and the same as now for existing BIO_read().

Comment (levitte) [inline crypto/bio/bio_meth.c]: Hmmm, re-reading what you wrote in the manual, I see that you intend 0 and 1 to indicate if any data was successfully read or not, not necessarely if there was an error (i.e. a code was addded to the error queue) or not.  So ok, I accept your reasoning.

**Comment by mattcaswell** [comment]
BIO_read()/BIO_write() return types are a nightmare!! There are all sorts of inconsistencies and problems.

BIO_read()/BIO_write() say this about their return type:

````
All these functions return either the amount of data successfully read or
written (if the return value is positive) or that no data was successfully
read or written if the result is 0 or -1. If the return value is -2 then
the operation is not implemented in the specific BIO type.  The trailing
NUL is not included in the length returned by BIO_gets().

=head1 NOTES

A 0 or -1 return is not necessarily an indication of an error. In
particular when the source/sink is non-blocking or of a certain type
it may merely be an indication that no data is currently available and that
the application should retry the operation later.
````

Some variation of that wording has been there since the docs were first written in 2000. However it seems that, despite that, implemented code makes different assumptions about how it works. If libssl gets an <=0 return code from BIO_read() then it returns that same unchanged error code all the way back up the stack to the caller of SSL_read(). SSL_read() says this about its return type:

````
The following return values can occur:

=over 4

=item E<gt>0

The read operation was successful; the return value is the number of
bytes actually read from the TLS/SSL connection.

=item Z<>0

The read operation was not successful. The reason may either be a clean
shutdown due to a "close notify" alert sent by the peer (in which case
the SSL_RECEIVED_SHUTDOWN flag in the ssl shutdown state is set
(see L<SSL_shutdown(3)>,
L<SSL_set_shutdown(3)>). It is also possible, that
the peer simply shut down the underlying transport and the shutdown is
incomplete. Call SSL_get_error() with the return value B<ret> to find out,
whether an error occurred or the connection was shut down cleanly
(SSL_ERROR_ZERO_RETURN).

=item E<lt>0

The read operation was not successful, because either an error occurred
or action must be taken by the calling process. Call SSL_get_error() with the
return value B<ret> to find out the reason.
````

The <0 case for SSL_read() suggests it could be because of non-blocking IO ("...or action must be taken by the calling process."), whereas the 0 case doesn't seem to allow that possibility. However the reality is that 0 could mean that it is retryable if BIO_read() returned 0.

We have test code that looks at the return code of SSL_read() and fails outright if a 0 is returned, but allows the possibility of a retry if -1 is returned. See handshake_helper.c:
````
        ret = SSL_read(peer->ssl, peer->read_buf, peer->read_buf_len);
        if (ret > 0) {
            TEST_check(ret <= peer->bytes_to_read);
            peer->bytes_to_read -= ret;
        } else if (ret == 0) {
            peer->status = PEER_ERROR;
            return;
        } else {
            int error = SSL_get_error(peer->ssl, ret);
            if (error != SSL_ERROR_WANT_READ) {
                peer->status = PEER_ERROR;
                return;
            } /* Else continue with write. */
        }
````
There are similar assumptions that a 0 return from SSL_read() is a fatal (not-retryable) error in ssltest_old.c

On the other hand we have other code which assumes stuff about the implementation of the underlying BIO. A mem BIO will return 0 if it runs out of data. We have code that knows that it is using a mem BIO and looks for the 0 return value as an indicator for it to finish. E.g. see bio_enc_test.c:

````
for (len = 0; (delta = BIO_read(b, out + len, i)); ) {
    len += delta;
}
````

In crypto/cms/cms_smime.c there are other assumptions about the underlying BIO:
````
        i = BIO_read(in, buf, sizeof(buf));
        if (i <= 0) {
            if (BIO_method_type(in) == BIO_TYPE_CIPHER) {
                if (!BIO_get_cipher_status(in))
                    goto err;
            }
            if (i < 0)
                goto err;
            break;
        }
````
Here it is known we are working with a mem BIO, so we look for a 0 return as an indicator of running out of data (which is ok), while -1 is a non-recoverable failure!

In other words, depending on the context sometimes we treat 0 as recoverable and -ve as non-recoverable; sometimes we treat it the other way around so that 0 is non-recoverable and -ve is recoverable; and the docs say that either could be recoverable!!!

File: crypto/bio/bio_meth.c (hunk 2 above)
Comment (mattcaswell) [inline crypto/bio/bio_meth.c]: Fixed.

**Comment by mattcaswell** [comment]
New commits pushed that address some travis failure issues; ensure that BIO_read_ex() and BIO_write_ex() only ever return 0 or 1 as per the docs; and to address the INT_MAX issue raised above.

Ping @levitte.

File: crypto/bio/bio_lib.c [hunk 4]
```diff
@@ -174,124 +244,224 @@ int BIO_method_type(const BIO *b)
...
-    i = b->method->bputs(b, in);
+    if (b->callback != NULL || b->callback_ex != NULL)
+        ret = (int)bio_call_callback(b, BIO_CB_PUTS | BIO_CB_RETURN, in, 0, 0,
+                                     0L, ret, &written);
 
-    if (i > 0)
-        b->num_write += (uint64_t)i;
+    if (ret > 0) {
+        if (written > INT_MAX)
+            ret = -1;
```
Comment (levitte) [inline crypto/bio/bio_lib.c]: *ouch*

I know no way around this, but yeah, this is problematic.

(not expecting a fix)

File: crypto/bio/bio_lib.c [hunk 5]
```diff
@@ -174,124 +244,224 @@ int BIO_method_type(const BIO *b)
...
+    if (b->callback != NULL || b->callback_ex != NULL)
+        ret = (int)bio_call_callback(b, BIO_CB_GETS | BIO_CB_RETURN, out, outl,
+                                     0, 0L, ret, &read);
 
-    if (cb != NULL)
-        i = (int)cb(b, BIO_CB_GETS | BIO_CB_RETURN, in, inl, 0L, (long)i);
-    return (i);
+    if (ret > 0) {
+        if (read > INT_MAX)
+            ret = -1;
```
Comment (levitte) [inline crypto/bio/bio_lib.c]: This should be an impossible condition, since `outl` is an int...  If we get here, there's something deeply wrong with the underlying implementation.  This is a place where I would actually `assert(read <= INT_MAX)`

File: crypto/bio/bio_meth.c [hunk 6]
```diff
@@ -51,24 +51,93 @@ void BIO_meth_free(BIO_METHOD *biom)
...
     return biom->bwrite;
 }
 
+/* Conversion for old style bwrite to new style */
+int bwrite_conv(BIO *bio, const char *in, size_t inl, size_t *written)
+{
+    int ret;
+
+    if (inl > INT_MAX)
+        return 0;
```
Comment (levitte) [inline crypto/bio/bio_meth.c]: Same as we had in `BIO_read_ex()` earlier, I'd rather do `inl = INT_MAX;` here.

File: crypto/bio/bio_lib.c [hunk 7]
```diff
@@ -174,124 +244,224 @@ int BIO_method_type(const BIO *b)
...
+    }
 
-    if (cb != NULL)
-        i = (int)cb(b, BIO_CB_PUTS | BIO_CB_RETURN, in, 0, 0L, (long)i);
-    return (i);
+    return ret;
 }
 
-int BIO_gets(BIO *b, char *in, int inl)
+int BIO_gets(BIO *b, char *out, int outl)
```
Comment (levitte) [inline crypto/bio/bio_lib.c]: This parameter rename is confusing to me.  I *write* things *out* and I *read* things *in*.

Comment (levitte) [inline crypto/bio/bio_lib.c]: ... but I won't hold it against you.  I just noticed that you're following an already existing pattern.

File: crypto/bio/bio_meth.c (hunk 6 above)
Comment (levitte) [inline crypto/bio/bio_meth.c]: ... sorry, not `BIO_read_ex()`, but `bread_conv()`

File: ssl/bio_ssl.c [hunk 8]
```diff
@@ -171,9 +179,9 @@ static int ssl_write(BIO *b, const char *out, int outl)
 
     BIO_clear_retry_flags(b);
 
-    /*
-     * ret=SSL_do_handshake(ssl); if (ret > 0)
-     */
+    if (outl > INT_MAX)
+        return 0;
```
Comment (levitte) [inline ssl/bio_ssl.c]:
``` C
        outl = INT_MAX;
```

File: include/openssl/bio.h [hunk 9]
```diff
@@ -545,8 +552,10 @@ int BIO_get_shutdown(BIO *a);
 void BIO_vfree(BIO *a);
 int BIO_up_ref(BIO *a);
 int BIO_read(BIO *b, void *data, int len);
+int BIO_read_ex(BIO *b, void *out, size_t outl, size_t *read);
```
Comment (levitte) [inline include/openssl/bio.h]: Ok, sorry for being wishy washy on the subject, but...  would you mind renaming the parameters here as follows?

``` C
int BIO_read_ex(BIO *b, void *data, size_t datal, size_t *read);
```

File: include/openssl/bio.h [hunk 10]
```diff
@@ -545,8 +552,10 @@ int BIO_get_shutdown(BIO *a);
...
 int BIO_gets(BIO *bp, char *buf, int size);
 int BIO_write(BIO *b, const void *data, int len);
+int BIO_write_ex(BIO *b, const void *in, size_t inl, size_t *written);
```
Comment (levitte) [inline include/openssl/bio.h]: ... and likewise here:

``` C
int BIO_write_ex(BIO *b, const void *data, size_t datal, size_t *written);
```

File: ssl/bio_ssl.c [hunk 11]
```diff
@@ -16,8 +16,8 @@
 #include <openssl/err.h>
 #include "ssl_locl.h"
 
-static int ssl_write(BIO *h, const char *buf, int num);
-static int ssl_read(BIO *h, char *buf, int size);
+static int ssl_write(BIO *h, const char *buf, size_t num, size_t *written);
+static int ssl_read(BIO *b, char *out, size_t outl, size_t *read);
```
Comment (levitte) [inline ssl/bio_ssl.c]: Change to this:

``` C
static int ssl_read(BIO *b, char *buf, size_t size, size_t *read);
```

File: ssl/bio_ssl.c [hunk 12]
```diff
@@ -101,14 +103,19 @@ static int ssl_read(BIO *b, char *out, int outl)
 
     BIO_clear_retry_flags(b);
 
+    if (outl > INT_MAX)
+        return -1;
```
Comment (levitte) [inline ssl/bio_ssl.c]:
``` C
        outl = INT_MAX;
```

File: crypto/bio/bio_lib.c (hunk 5 above)
Comment (mattcaswell) [inline crypto/bio/bio_lib.c]: Well, yes and no. This could be an error outside of OpenSSL. The callback can be registered by application code, and could set `read > INT_MAX`. This is obviously an error in the application code but its probably not our place to assert things about application code. We *could* detect this in bio_call_callback I suppose...I'm not really sure what the correct course of action is there.

File: crypto/bio/bio_meth.c (hunk 6 above)
Comment (mattcaswell) [inline crypto/bio/bio_meth.c]: Done

File: ssl/bio_ssl.c (hunk 11 above)
Comment (mattcaswell) [inline ssl/bio_ssl.c]: Done

File: ssl/bio_ssl.c (hunk 12 above)
Comment (mattcaswell) [inline ssl/bio_ssl.c]: Done

File: ssl/bio_ssl.c (hunk 8 above)
Comment (mattcaswell) [inline ssl/bio_ssl.c]: Done

File: include/openssl/bio.h (hunk 10 above)
Comment (mattcaswell) [inline include/openssl/bio.h]: Done (as above).

**Comment by mattcaswell** [comment]
New commits pushed.

File: include/openssl/bio.h (hunk 9 above)
Comment (levitte) [inline include/openssl/bio.h]: It would be nice if the implementation matched the declaration...

File: crypto/bio/bio_lib.c (hunk 5 above)
Comment (levitte) [inline crypto/bio/bio_lib.c]: Ah...  yeah ok, if they do return something larger than `outl` (which can't be bigger than `INT_MAX`), then there really is an issue....  I would say that `BIO_gets()` should generate an error, then (especially since we return -1...  the method has most likely *not* set an error in this case).

File: include/internal/bio.h [hunk 13]
```diff
@@ -24,3 +26,8 @@ struct bio_method_st {
 
 void bio_free_ex_data(BIO *bio);
 void bio_cleanup(void);
+
+
+/* Old style to new style BIO_METHOD conversion functions */
+int bwrite_conv(BIO *bio, const char *in, size_t inl, size_t *written);
+int bread_conv(BIO *bio, char *out, size_t outl, size_t *read);
```
Comment (levitte) [inline include/internal/bio.h]: Similar to the public API changes, I would like to see `in` and `out` changed to `data`, and `inl` and `outl` to `datal`.

Comment (levitte) [inline include/internal/bio.h]: ... and the implementation changed similarly.

Comment (mattcaswell) [inline include/internal/bio.h]: Done

**Comment by mattcaswell** [comment]
Github is behaving badly this afternoon. I can't reply to one of your comments above for some reason. Anyway I fixed the BIO_gets() length check thing by adding a sanity check. I added a similar check for BIO_read*(). New commits pushed.

File: crypto/bio/bio_lib.c [hunk 14]
```diff
@@ -174,124 +244,235 @@ int BIO_method_type(const BIO *b)
...
+
+    if (b->callback != NULL || b->callback_ex != NULL)
+        ret = (int)bio_call_callback(b, BIO_CB_READ | BIO_CB_RETURN, data,
+                                     datal, 0, 0L, ret, read);
 
-    if (i > 0)
-        b->num_read += (uint64_t)i;
+    /* Shouldn't happen */
+    if (ret > 0 && *read > datal)
+        return -1;
```
Comment (levitte) [inline crypto/bio/bio_lib.c]: Please add an error code here.

File: crypto/bio/bio_lib.c [hunk 15]
```diff
@@ -174,124 +244,235 @@ int BIO_method_type(const BIO *b)
...
+        ret = 1;
+    }
+
+    if (b->callback != NULL || b->callback_ex != NULL)
+        ret = (int)bio_call_callback(b, BIO_CB_PUTS | BIO_CB_RETURN, in, 0, 0,
+                                     0L, ret, &written);
+
+    if (ret > 0) {
+        if (written > INT_MAX)
+            ret = -1;
```
Comment (levitte) [inline crypto/bio/bio_lib.c]: Please add an error code here

File: doc/crypto/BIO_meth_new.pod [hunk 16]
```diff
@@ -64,15 +74,23 @@ more information.
...
-called in response to the application calling BIO_read(). The parameters for the
-function have the same meaning as for BIO_read().
+BIO_meth_get_write_ex() and BIO_meth_set_write_ex() get and set the function
+used for writing arbitrary length data to the BIO respectively. This function
+will be called in response to the application calling BIO_write_ex() or
+BIO_write(). The parameters for the function have the same meaning as for
+BIO_write_ex(). Older code may call BIO_meth_get_write() and
+BIO_meth_set_write() instead. Applications should not call both
+BIO_meth_set_write_ex() and BIO_meth_set_write() or call BIO_meth_get_write()
+when the function was set with BIO_meth_set_write_ex().
```
Comment (kaduk) [inline doc/crypto/BIO_meth_new.pod]: Should we say that _ex takes precedence, as is done for the BIO callback?  I guess it's only for library-internal writes, so that's more of a messy thing to say...

File: crypto/bio/bio_lib.c [hunk 17]
```diff
@@ -174,124 +244,235 @@ int BIO_method_type(const BIO *b)
...
 }
 
-int BIO_read(BIO *b, void *out, int outl)
+/*
+ * This is essentially the same as BIO_read_ex() except that it allows
+ * 0 or a -ve value to indicate failure (retryable or not) in the return. This
+ * is for compatibility with the old style BIO_read(), where existing code may
+ * make assumptions about the return value that it might get.
+ */
+static int bio_read_intern(BIO *b, void *data, size_t datal, size_t *read)
```
Comment (kaduk) [inline crypto/bio/bio_lib.c]: I know @levitte explicitly asked for this format, but in some fonts, "datal" looks a lot like "data1".  "dlen" does not suffer from that issue (though probably has others of its own).  Anyway, I am not looking to start a bikeshed here, and would certainly be able to cope just fine with "datal".

File: crypto/bio/bio_lib.c [hunk 18]
```diff
@@ -13,6 +13,65 @@
...
+        if (inret && (oper & BIO_CB_RETURN)) {
+            if (*processed > INT_MAX)
+                return -1;
+            inret = *processed;
+        }
+    }
+
+    ret = b->callback(b, oper, argp, argi, argl, inret);
+
+    if (ret > LONG_MAX || ret < LONG_MIN)
```
Comment (kaduk) [inline crypto/bio/bio_lib.c]: How could either case happen if 'ret' is of type long?

File: crypto/bio/bio_lib.c (hunk 15 above)
Comment (mattcaswell) [inline crypto/bio/bio_lib.c]: Done

File: crypto/bio/bio_lib.c (hunk 18 above)
Comment (mattcaswell) [inline crypto/bio/bio_lib.c]: Good point. I'm not sure what I was thinking when I wrote that. Removed.

File: crypto/bio/bio_lib.c [hunk 19]
```diff
@@ -174,124 +244,235 @@ int BIO_method_type(const BIO *b)
     return b->method->type;
 }
 
-int BIO_read(BIO *b, void *out, int outl)
+/*
+ * This is essentially the same as BIO_read_ex() except that it allows
+ * 0 or a -ve value to indicate failure (retryable or not) in the return. This
```
Comment (mattcaswell) [inline crypto/bio/bio_lib.c]: Done

File: crypto/bio/bio_lib.c (hunk 17 above)
Comment (mattcaswell) [inline crypto/bio/bio_lib.c]: Renamed to `dlen`

File: doc/crypto/BIO_meth_new.pod (hunk 16 above)
Comment (mattcaswell) [inline doc/crypto/BIO_meth_new.pod]: Well we already say that you shouldn't set both an _ex and a non _ex, so I'm not sure we need to worry about this. From an API user perspective, they don't care what happens internally.

**Comment by mattcaswell** [comment]
New commit pushed addressing comments from @levitte and @kaduk.

**Comment by mattcaswell** [comment]
Rebased to get rid of the merge conflict warnings from github. Also resolved another shadowed variable warning, and fixed no-stdio builds. Ping @levitte.

**Comment by mattcaswell** [comment]
Pushed. Thanks.
